xwiki-rendering-api/src/main/java/org/xwiki/rendering/internal/parser/reference/GenericLinkReferenceParser.java

```diff
@@ -132,28 +132,28 @@
     @Override
     public ResourceReference parse(String rawReference)
     {
-        // Step 1: Check if it's a known URI by looking for one of the known URI schemes. If not, check if it's a URL.
+        StringBuilder content = new StringBuilder(rawReference);
         ResourceReference resourceReference = parseURILinks(rawReference);
-        if (resourceReference != null) {
-            return resourceReference;
-        }
 
+        // Step 1: Check if it's a known URI by looking for one of the known URI schemes. If not, check if it's a URL.
+        if (parseURILinks(rawReference) != null) {
+            resourceReference = parseURILinks(rawReference);
+        }
         // Step 2: Look for an InterWiki link
-        StringBuilder content = new StringBuilder(rawReference);
-        resourceReference = parseInterWikiLinks(content);
-        if (resourceReference != null) {
-            return resourceReference;
+        else if (parseInterWikiLinks(content) != null) {
+            resourceReference = parseInterWikiLinks(content);
```

Comment:
This is very bad for performances, you are re-executing several time the same code.

Reply:
Didn’t see that, will fix. Thanks.

Reply:
You actually did this mistake in most of your modifications (parseURILinks is now executed 3 times now).

Reply:
As of the latest commit, the repeated executions should all be fixed. Thanks!